File: swift/03p2.swift
import Foundation

func getSingleClaimId (patches: [[[Int]]], claimIds: [Int]) -> Int {
  var ids = Set(claimIds)
  for row in patches {
    for c in row {
      if c.count > 1 {
        for claimId in c { ids.remove(claimId) }
      }
    }
  }
  return ids.removeFirst()
}

var claims: [String] = [];
while let claim = readLine() { claims.append(claim) }

let ints: [Int] = []
var patches = Array(repeating: Array(repeating: ints, count: 1000),
                    count: 1000)
var claimIds: [Int] = []
for claim in claims {
  var claimEls = claim.components(separatedBy: " @ ")
  claimEls[0].removeFirst(1)
  let claimId = Int(claimEls[0]) ?? 0
  claimIds.append(claimId)
  let claimSpec = claimEls[1]
  let specGeoEls = claimSpec.components(separatedBy: ": ")
  let startCoord = specGeoEls[0]
  let area = specGeoEls[1]
  let startCoordEls = startCoord.components(separatedBy: ",")
  let startX = Int(startCoordEls[0]) ?? 0
  let startY = Int(startCoordEls[1]) ?? 0
  let areaEls = area.components(separatedBy: "x")
  let endX = startX + (Int(areaEls[0]) ?? 0)
  let endY = startY + (Int(areaEls[1]) ?? 0)

  for y in startY..<endY {
    for x in startX..<endX { patches[y][x].append(claimId) }
  }
}

let claimId = getSingleClaimId(patches: patches, claimIds: claimIds)
print("\(claimId)")

File: python/03p2.py
#!/usr/bin/env python

import sys

def get_single_claim_id(patches, claim_ids):
    ids = set(claim_ids)
    for cols in patches:
        for cell_ids in cols:
            if len(cell_ids) > 1:
                for claim_id in cell_ids: ids.discard(claim_id)
    return ids.pop()

claims = [claim.rstrip() for claim in sys.stdin]
SQUARE_WIDTH = 1000
patches = [[[] for x in range(SQUARE_WIDTH)] for y in range(SQUARE_WIDTH)]
claim_ids = []
for claim in claims:
    claim_els = claim.split(" @ ")
    claim_id = int(claim_els[0][1:] or 0)
    claim_ids.append(claim_id)
    claim_spec = claim_els[1]
    spec_geo_els = claim_spec.split(": ")
    start_coord = spec_geo_els[0]
    area = spec_geo_els[1]
    start_coord_els = start_coord.split(",")
    start_x = int(start_coord_els[0]) or 0
    start_y = int(start_coord_els[1]) or 0
    area_els = area.split("x")
    end_x = int(start_x) + (int(area_els[0]) or 0)
    end_y = int(start_y) + (int(area_els[1]) or 0)

    for y in range(start_y, end_y):
        for x in range(start_x, end_x): patches[y][x].append(claim_id)

claim_id = get_single_claim_id(patches, claim_ids)
print(claim_id)
